### PR TITLE
Enable strict null checks

### PR DESCRIPTION
**Pull request by syimyuzya**
Annotations of nullable properties and parameters now finally appear in the docs!

Also enabled non-null assertions for the sake of readability and maintainability.

## Note about ESLint's `no-non-null-assertion`

(copied from comments in .eslintrc.json)

Non-null assertions enabled for the sake of readability and maintainability.

Since `x!` is just equivalent to `x as NonNull<typeof x>`, and using `as` is even less safe than `!` (as `as` allows arbitrary type assertion), `!` is actually the best practice to assert something is not null when the type system is not smart enough (or when it requires some kind of "type-gymnastics", which only makes the code more delicate and actually hinders readability).

However, it is still recommended that use of non-null assertions should still be limited to these cases (use `as` or comment about it otherwise):

1. The assertion is clearly confirmable by simply looking at statements just before it.
  ```js
  if (!map.has(k)) map.set(k, v); // This line ensures the next `get` to be non-null
  map.get(k)!
  ```

2. An accidental null value can cause the program to fail fast (thus easy to test and fix).
  ```js
  let 等: string;
  if (... ) 等 = ...;
  else if (...) 等 = ...;
  return new 音韻地位(..., 等!, ...); // an `undefined` will surely trigger an error in the constructor
  ```

3. A type-sound version (if possible) would require too much type-gymnastics

4. `null` is actually an intended value
  ```js
  所有呼.indexOf(呼) + 1 // `indexOf` returns -1 if 呼 is null, which is intended
  [...'幫見影'].includes(組!) // returns false if 組 is null, which is intended
  ```

### CODE REVIEW

**Comment by graphemecluster** [comment]
###### 如果 tsconfig.json（還有 test 配套）可以跟 tshet-uinh-deriver-tools 保持一致就好了（小聲）

**Comment by syimyuzya** [comment]
> ###### 如果 tsconfig.json（還有 test 配套）可以跟 tshet-uinh-deriver-tools 保持一致就好了（小聲）

跟 tshet-uinh-deriver-tools 一致的話要求還要更高些（比如 Node 16+，因為 `Error.cause` 是 ES2022）ww

（但我覺得可以 (?)

**Comment by graphemecluster** [comment]
一直忘了 submit review w

File: src/lib/音韻地位.ts [hunk 1]
```diff
@@ -780,7 +784,9 @@ export class 音韻地位 {
    * 'p'
    * ```
    */
-  判斷<T, E = undefined>(規則: 判斷規則列表<T>, error?: E, fallback?: boolean): E extends Falsy ? T | null : T {
+  判斷<T>(規則: 判斷規則列表<T>, error?: null, fallback?: boolean): T | null;
+  判斷<T, E>(規則: 判斷規則列表<T>, error: E, fallback?: boolean): T;
+  判斷<T, E = undefined>(規則: 判斷規則列表<T>, error?: E, fallback?: boolean): T | null {
```
[the remark this answers is not in the document]
Comment (syimyuzya) [reply src/lib/音韻地位.ts:789]: 咦好耶～

[the remark this answers is not in the document]
Comment (syimyuzya) [reply src/lib/音韻地位.ts:789]: > 欸也可以這樣寫

突然意識到只有這最後一版寫法是合適的，更上面的寫法（以及我一開始的寫法）都是有問題的🙈

因為返回類型 `T` 是 `T | null` 的子集（對應的參數可能取值自然也是子集），它表示更強的靜態分析條件，應該放前面優先匹配。

所以聲明可以這樣寫：

（取消 `null` 恢復用 boolean，空串視作拋異常）

```ts
判斷<T>(規則: 判斷規則列表<T>, error: string | true, fallback?: boolean): T;
判斷<T>(規則: 判斷規則列表<T>, error?: string | boolean, fallback?: boolean): T | null;
```

【PS 空串也拋異常可以讓靜態分析條件與類型（而非真值）更好地對應起來

[the remark this answers is not in the document]
Comment (syimyuzya) [reply src/lib/音韻地位.ts]: f9ccba6 現在是改成了這樣w

**Pull request merged**